Log results to platform specific location

nickwanninger opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: Logging historical data to $XDG_DATA_HOME/thokr.log

<!-- Why are these changes necessary? -->

**Why**: So nerds like me can plot progress over time

<!-- How were these changes implemented? -->

**How**: Results are logged whenever results are calculated. It may be smart to add a configuration to disable this, but that could be another comit 

<!-- Have you done all of these things?  -->

**Checklist**: N/A

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Documentation added
- [ ] Tests added
- [x] No linting errors / broken tests
- [x] Merge ready
      <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

<!-- feel free to add additional comments -->

coloradocolby commented on src/thok.rs:
```diff
@@ -141,6 +171,8 @@ impl Thok {
         }
         self.accuracy = ((correct_chars.len() as f64 / self.input.len() as f64) * 100.0).round();
+
+        let _ = self.save_results();
```
Any reason for the `let _`?

nickwanninger replied:
It's the least gross way to ignore an error. It could be replaced with `.ok()` or `.drop()`.

I think it makes sense to ignore the error here, as failing to log results shouldn't crash the program. Maybe instead it should log to stderr or display on the results screen?

coloradocolby commented on src/thok.rs:
```diff
@@ -68,6 +71,33 @@ impl Thok {
     }
 
+    pub fn save_results(&self) -> io::Result<()> {
+        let log_path = dirs::data_dir().unwrap().join("thokr.log");
```
Also I think I'd rather thokr be a folder and this file be nested inside, that way there's already a folder for potential more files down the road

coloradocolby replied:
Just checking that there should be no issues with `dirs` and windows? I'm extremely unfamiliar with windows os lol. Do they have a `XDG_DATA_HOME`?

nickwanninger replied:
Yeah, I agree with the folder idea. `$XDG_DATA_HOME/thokr/log.csv`?

Also yeah I think that dirs is good on windows (at least they say so haha)

nickwanninger replied:
Pushed this change

coloradocolby commented on src/thok.rs:
```diff
@@ -68,6 +71,33 @@ impl Thok {
+            .open(log_path)?;
+
+        if needs_header {
+            writeln!(log_file, "# Date, WPM, Accuracy, Standard Deviation")?;
```
Not a big fan of `# Date` -- mostly just the `#` feels confusing. I'd want to add a few more headers here too, namely to track what parameters the test was run with. Was it 15 words? 100? Was it a 30 second test? 60? Stuff like that!

nickwanninger replied:
I'm happy to add that for sure.  I'm not the most familiar with the project, so what might be the best way to get the word count other than counting the `' '` chars in `self.input`?

coloradocolby commented on src/thok.rs:
```diff
@@ -68,6 +71,33 @@ impl Thok {
+        writeln!(
+            log_file,
+            "{}, {}, {}, {}",
+            Utc::now(),
```
```suggestion
            Local::now(),
```
since this is user facing I think local time would be more convenient

coloradocolby reviewed (commented):
Left a few comments, overall a big fan of the idea!

nickwanninger commented:
It should now log with the wordcount and duration. I modified the Thok structure to include a test_duration, and renamed the existing `duration` to `time_remaining` to clarify things a bit

coloradocolby commented:
Will cut a release in a bit, but merging for now! Thanks for the hard work @nickwanninger !

nickwanninger commented:
For sure! Happy to help!
